File: leetcode-algorithms/594. Longest Harmonious Subsequenc/Solution.cpp
// We define a harmonious array is an array where the difference between its maximum value and its minimum value is exactly 1.
//
// Now, given an integer array, you need to find the length of its longest harmonious subsequence among all its possible subsequences.
//
// Example 1:
// Input: [1,3,2,2,5,2,3,7]
// Output: 5
// Explanation: The longest harmonious subsequence is [3,2,2,2,3].
// Note: The length of the input array will not exceed 20,000.
class Solution {
public:
    int findLHS(vector<int>& nums) {
        set<int> se;
        map<int, int> countmap;

        for (auto v : nums) {
            countmap[v]++;
            se.insert(v);
        }

        int maxlen = 0;
        for (auto v : se) {
            if (countmap.count(v + 1)) maxlen = max(maxlen, countmap[v] + countmap[v + 1]);
            if (countmap.count(v - 1)) maxlen = max(maxlen, countmap[v] + countmap[v - 1]);
        }

        return maxlen;

    }
};

File: leetcode-algorithms/594. Longest Harmonious Subsequenc/Solution.java
class Solution {
    public int findLHS(int[] nums) {
    	Set<Integer> vset = new HashSet<>();
    	for (int v : nums) vset.add(v);

    	int maxl = 0;
        for (int i = 0; i < nums.length; i++) {
        	maxl = Math.max(maxl, find(nums, i, vset));
        }

        return maxl;
    }

    private int find(int[] nums, int idx, Set<Integer> vset) {
    	int v = nums[idx];
    	int up = 0;
    	if (vset.contains(v + 1)) {
    		for (int i = 0; i < nums.length; i++) {
    			if ((nums[i] == v) || (nums[i] == v + 1)) up++;
    		}
    	}

    	int down = 0;
    	if (vset.contains(v - 1)) {
    		for (int i = 0; i < nums.length; i++) {
    			if ((nums[i] == v) || (nums[i] == v - 1)) down++;
    		}
    	}

    	return Math.max(down, up);
    }
}
